### PR TITLE
Sphinx: fail when a command fails + other stuff

### LINKED ISSUE #7755: How to ensure that the Coq snippets in the manual are not broken?
It has been a long standing problem that the Coq examples in the manual can sometimes become outdated without us noticing; they can even start failing suddenly. The problem was there with coq-tex and still exists in the Sphinx based manual.

A solution that has been proposed and never implemented is to forbid failures (cc @silene @maximedenes). There are a few places where we want to show Coq failing, but it could still be done using `Fail` (although the output with `Fail` is less advanced, in particular it doesn't show the failure location).

Yet, I fear that this mechanism would still be too simple to catch a lot of problems: when we describe the supposed output in terms more complex than "this works" or "this fails". That's why I'm currently considering that we should instead have a system that is closer to how the test-suite `output/` system works, especially after @SkySkimmer's improvement in #7663. We would store the output along side the input. We could even use the stored outputs by default instead of regenerating them (it would allow the `sphinx` target not to depend on `world`) and have a distinct target (included in `test-suite`) that would regenerate outputs and check them against what was stored (plus an additional target to "approve" new outputs after careful review, also useful when we add a new example).

### LINKED ISSUE #9407: Manual examples broken with nested proofs
Code snippets like here:

https://coq.inria.fr/distrib/current/refman/language/gallina-extensions.html?highlight=generalizable#implicit-generalization

trigger an error about nested proofs. It's a pity that errors on these examples (which are meant to be successfully executed) is not caught by our documentation build.

### PR DESCRIPTION
**Pull request by SkySkimmer**
This uses a new coqtop-only option "Coqtop Fast Death", not sure where to put the doc for it.

Having it as an option allows us to disable it for a block with `coqtop:: fail`, but that's very little used so if needed it could become a command line flag. Those are heavier to implement though.

Closes #7755
Closes #9407.

Some notes:
- Undo is quite broken, eg `Definition x := 1. Undo 1.` says "cannot undo". It seems to work OK to undo tactics but not regular commands. Maybe sphinx should use Back instead? OTOH Back will see a lemma with 42 tactics as just 1 thing so sphinx basing the number on how many commands were sent is still broken. Can we come up with some more structured state reset system?
- ssreflect was using `coqtop:: in` instead of `coqdoc::` quite a lot
- there's a syntax error in ssreflect which seems like it's supposed to work, see FIXME in the rst (I turned it into coqdoc until someone looks at it), cc @gares 
- when a command fails coqtop does `exit 1` then sphinx fails with an exception, but it can be difficult to identify where the rst is broken (the exception doesn't mention the line number, it only says the last command sent to coqtop which can be something common like `Abort` and the last 100 characters of the error message). I don't know how to improve that.
- when a command failed, I fixed it and reran sphinx, it re-processed all previous files including successful ones wasting quite some time. I ended up cheating by doing `ln current-broken-file.rst canonical-structures.rst` (canonical-structures is the first to get processed) but is there a better way?
- sphinx wanted more recursion depth (breaking while processing ring.rst), so I doubled it. No clue what's going on.

Depends https://github.com/coq/coq/pull/9553

### CODE REVIEW

**Comment by SkySkimmer** [comment]
>Some notes:

Almost forgot:
- the `coqtop::` in the micromega chapter fails due to lack of csdp, we should consider installing it on the CI images

**Review by ejgallego — CHANGES_REQUESTED** [review]
Umm, I am not sure I like the option [which should be renamed anyways]

I suggest you instead use the `-l` option of `coqtop` if you can.

**Comment by SkySkimmer** [comment]
>I suggest you instead use the -l option of coqtop if you can.

How would that work with the sphinx model? It's basically intended for interactive coqtop.

**Comment by Zimmi48** [comment]
First, thanks for working on this. I would personally have preferred a solution comparing the output with a saved output and a promoting functionality, but what you propose is already a serious improvement on the current situation and solves the main issue.

Could you move the fixes to the manual to a separate PR so that they can be backported to 8.9?

>Undo is quite broken, eg `Definition x := 1. Undo 1.` says "cannot undo". It seems to work OK to undo tactics but not regular commands. Maybe sphinx should use Back instead? OTOH Back will see a lemma with 42 tactics as just 1 thing so sphinx basing the number on how many commands were sent is still broken. Can we come up with some more structured state reset system?

See the discussion in #7353.

>when a command fails coqtop does `exit 1` then sphinx fails with an exception, but it can be difficult to identify where the rst is broken (the exception doesn't mention the line number, it only says the last command sent to coqtop which can be something common like `Abort` and the last 100 characters of the error message). I don't know how to improve that.

@cpitclaudel might know...

>when a command failed, I fixed it and reran sphinx, it re-processed all previous files including successful ones wasting quite some time. I ended up cheating by doing `ln current-broken-file.rst canonical-structures.rst` (canonical-structures is the first to get processed) but is there a better way?

Unfortunately, Sphinx's caching is terrible. Both unsound and too weak.

**Comment by ejgallego** [comment]
> How would that work with the sphinx model? It's basically intended for interactive coqtop.

`-l` basically simulates the interactive loading of a file, but however you are right, goal printing was removed from the loop in `Vernac` , let me think....

**Comment by cpitclaudel** [comment]
Thanks for working on this, this is great work. I'll look at the details of the questions a bit later today (thanks for the ping also @Zimmi48)

**Review by ejgallego — APPROVED** [review]
Ok then with the option, but please rename to something closer to "Coqtop Exit On Error"

File: toplevel/coqloop.ml [hunk 1]
```diff
@@ -410,7 +415,8 @@ let rec vernac_loop ~state =
     let loc = Loc.get_loc info in
     let msg = CErrors.iprint (e, info) in
     TopErr.print_error_for_buffer ?loc Feedback.Error msg top_buffer;
-    vernac_loop ~state
+    if fast_death () then exit 1
+    else vernac_loop ~state
```
Comment (ejgallego) [inline toplevel/coqloop.ml]: I'd prefer:
```
if exit_on_error () then exit 1;
vernac_loop ~state
```
as to make the critical importance of tail-rec clear here.

Comment (SkySkimmer) [reply toplevel/coqloop.ml]: As you like.

Comment (ejgallego) [reply toplevel/coqloop.ml]: Thanks

**Comment by gares** [comment]
>     * Undo is quite broken,

It is not broken, it is what Hugo wanted it to do.

File: doc/sphinx/proof-engine/ssreflect-proof-language.rst [hunk 2]
```diff
@@ -2997,7 +2996,7 @@ illustrated in the following example.
 
   .. coqtop:: none
 
-     Abort All.
+     Abort All. Reset Test.
```
Comment (gares) [inline doc/sphinx/proof-engine/ssreflect-proof-language.rst]: No, please, don't use Reset other than `Initial`

Comment (SkySkimmer) [reply doc/sphinx/proof-engine/ssreflect-proof-language.rst]: OK but why?

**Comment by SkySkimmer** [comment]
>there's a syntax error in ssreflect which seems like it's supposed to work, see FIXME in the rst (I turned it into coqdoc until someone looks at it), cc @gares

WG: it isn't intended to be executed, leaving as `coqdoc::`

File: doc/sphinx/proof-engine/ssreflect-proof-language.rst (hunk 2 above)
Comment (gares) [reply doc/sphinx/proof-engine/ssreflect-proof-language.rst]: Sorry, it was a reflex. If this is for coqtop, then it is fine

**Comment by cpitclaudel** [comment]
> when a command failed, I fixed it and reran sphinx, it re-processed all previous files including successful ones wasting quite some time. I ended up cheating by doing `ln current-broken-file.rst canonical-structures.rst` (canonical-structures is the first to get processed) but is there a better way?

I opened a Sphinx bug to understand why caches are discarded upon any error (https://github.com/sphinx-doc/sphinx/issues/6035)

> when a command fails coqtop does `exit 1` then sphinx fails with an exception, but it can be difficult to identify where the rst is broken (the exception doesn't mention the line number, it only says the last command sent to coqtop which can be something common like `Abort` and the last 100 characters of the error message). I don't know how to improve that.

Here's a patch that should fix this: https://gist.github.com/cpitclaudel/8a9c40ff54485a240d063b02ad9d00f7 . I think it makes sense to merge it as part of this PR, but I can also submit it as a separate PR if preferred.  Now the output looks like this:

```
Extension error:
/build/coq/master/doc/sphinx/addendum/canonical-structures.rst:39: Error while sending the following sentence to coqtop: Module EQ.
```

**Comment by SkySkimmer** [comment]
>Now the output looks like this:

Can we get it to include Coq's error message?

**Comment by cpitclaudel** [comment]
> Can we get it to include Coq's error message?

I've updated the patch at https://gist.github.com/cpitclaudel/8a9c40ff54485a240d063b02ad9d00f7. Can you see if that helps?  It's tricky to capture the message after Coq crashes.

**Comment by SkySkimmer** [comment]
> It's tricky to capture the message after Coq crashes.

See new commit, seems to work.

**Comment by Zimmi48** [comment]
@SkySkimmer Would you mind using the "Co-authored-by" syntax for crediting Clément instead of the @-mention in the commit message?

**Comment by SkySkimmer** [comment]
>@SkySkimmer Would you mind using the "Co-authored-by" syntax for crediting Clément instead of the @-mention in the commit message?

Like this?

**Comment by SkySkimmer** [comment]
>Could you move the fixes to the manual to a separate PR so that they can be backported to 8.9?

https://github.com/coq/coq/pull/9553

**Comment by Zimmi48** [comment]
>Like this?

Like this:
```
Co-authored-by: Clément Pit-Claudel <clement.pitclaudel@live.com>
```

File: doc/tools/coqrst/coqdomain.py [hunk 3]
```diff
@@ -866,36 +868,58 @@ def make_rawsource(pairs, opt_input, opt_output):
...
+        for sentence in self.split_sentences(node.rawsource):
+            pairs.append((sentence, repl.sendone(sentence)))
+        if opt_fail:
+            repl.sendone("Set Coqtop Exit On Error.")
+        if opt_undo:
+            repl.sendone("Undo {}.".format(len(pairs) + 2 * opt_fail))
+
+        dli = nodes.definition_list_item()
+        for sentence, output in pairs:
+            # Use Coqdoq to highlight input
```
Comment (Zimmi48) [inline doc/tools/coqrst/coqdomain.py]:
```suggestion
            # Use Coqdoc to highlight input
```

**Review by Zimmi48 — COMMENTED** [review]
Could you please also remove the `undo` option in this PR? This would avoid me requesting once more than @vbgl rebases his commit after I merge this PR. Cf. c1ad06146504a12653fa771bdc9e5f4a5bb46cba.

Could you document the `Coqtop Exit on Error` option in the manual as an internal-only option? (The best location is likely in https://coq.inria.fr/refman/practical-tools/coq-commands.html#interactive-use-coqtop.)

While you are at it (and only if you don't mind), I'd like to add two more control options:
- `abort`, which runs `Abort All.` at the end of the block;
- `restart` which runs `Restarts` at the end of the block.

**Comment by SkySkimmer** [comment]
>Could you please also remove the undo option in this PR? This would avoid me requesting once more than @vbgl rebases his commit after I merge this PR.

That got merged right?

>Could you document the Coqtop Exit on Error option in the manual as an internal-only option? 

Done.

>Coqdoq

Done.

>While you are at it (and only if you don't mind), I'd like to add two more control options:

Maybe some other time.

**Comment by vbgl** [comment]
No, the PR that was merged did not include the removal of the `undo` option. It only included changes to the contents of the manual.

**Comment by SkySkimmer** [comment]
>No, the PR that was merged did not include the removal of the undo option

OK, removed undo.

**Review by Zimmi48 — APPROVED** [review]

**Comment by Zimmi48** [comment]
It looks like there is still an unexpected failure in the refman...

**Comment by Zimmi48** [comment]
FTR I have fixed the error and intend to force-push to this branch.

**Comment by SkySkimmer** [comment]
OK thanks

**Comment by Zimmi48** [comment]
In addition to fixing the error, I used this PR to push additional commits adding support for the `abort` and `restart` options to the `coqtop` directive, I used these options throughout the manual, and I refactored (in a subsequent commit) the option system of the `coqtop` directive. In particular, in this last commit, I made the display option mandatory: before, it defaulted to `none` if a non-display option was passed and to `in` if no option was passed, and it was not documented at all.

@cpitclaudel Given what I just pushed, I removed my assignment and would appreciate if you could review and merge the PR.

**Comment by Zimmi48** [comment]
Ideally, we would also like the `fail` option to raise a failure if the `coqtop` block did not fail but I don't think this can be done in the current setting (and it's not used often enough to be that important).

**Comment by Zimmi48** [comment]
This has now a trivial conflict with master. @SkySkimmer do you want to fix it or shall I do it?

File: doc/tools/coqrst/coqdomain.py [hunk 4]
```diff
@@ -827,22 +827,41 @@ def split_sentences(source):
...
+        options = node['coqtop_options']
+
+        # Behavior options
         opt_reset = 'reset' in options
-        opt_all, opt_none = 'all' in options, 'none' in options
-        opt_input, opt_output = opt_all or 'in' in options, opt_all or 'out' in options
+        opt_fail = 'fail' in options
+        opt_restart = 'restart' in options
+        opt_abort = 'abort' in options
+        options = options - set(('reset', 'fail', 'restart', 'abort'))
```
Comment (cpitclaudel) [inline doc/tools/coqrst/coqdomain.py]: Sets are usually written as `{…}` instead of `set((…))`

File: doc/tools/coqrst/coqdomain.py [hunk 5]
```diff
@@ -827,22 +827,41 @@ def split_sentences(source):
...
 
-        unexpected_options = list(set(options) - set(('reset', 'undo', 'all', 'none', 'in', 'out')))
+        unexpected_options = list(options - set(('all', 'none', 'in', 'out')))
```
Comment (cpitclaudel) [inline doc/tools/coqrst/coqdomain.py]: Same here

**Review by cpitclaudel — APPROVED** [review]
LGTM. I will merge when the conflict is resolved.

**Comment by cpitclaudel** [comment]
Oh, woops; the conflict is already resolved.

File: doc/tools/coqrst/coqdomain.py (hunk 4 above)
Comment (Zimmi48) [reply doc/tools/coqrst/coqdomain.py]: Sorry, I'm not an advanced Pythonista. I only copied what was already there.

**Pull request merged**